aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu

```diff
@@ -19,11 +19,57 @@
 #include <thrust/transform.h>
 #include <thrust/unique.h>
 #include <thrust/system/cuda/execution_policy.h>
+#include <thrust/binary_search.h>
 #include <c10/macros/Macros.h>
 
 namespace at { namespace native {
 
 using namespace at::sparse;
+using at::cuda::detail::TensorInfo;
+using at::cuda::detail::getTensorInfo;
+
+namespace {
+
+template <typename scalar_t>
+#ifdef __HIP_PLATFORM_HCC__
+C10_LAUNCH_BOUNDS_1(512)
+#endif
+__global__ void _sparse_mask_copy_kernel(
+  int64_t total_threads,
+  int64_t t_nnz,
+  const TensorInfo<int64_t, int64_t> t_indices_ti,
+  const TensorInfo<int64_t, int64_t> mask_indices_ti,
+  const TensorInfo<int64_t, int64_t> t_indices_pos_ti,
+  const TensorInfo<scalar_t, int64_t> t_values_ti,
+  TensorInfo<scalar_t, int64_t> r_values_ti
+) {
+  const int64_t i = blockIdx.x * blockDim.x + threadIdx.x;
+  if (i >= total_threads) return;
+  const int64_t j = t_indices_pos_ti.data[i];
+
+  bool has_match = false;
+  if (j >= 0 &&  j < t_nnz && t_indices_ti.data[j] == mask_indices_ti.data[i]) {
+    has_match = true;
+  }
+
+  int64_t values_stride0 = r_values_ti.strides[0];
```

Comment:
How does this work with multidimensional dense tensors?

Reply:
this this works with multidimensional sparse tensors like:

```python
def test_hybrid():
    indices = torch.LongTensor([[2, 3, 0], 
                                [7, 9, 0]])
    x = torch.tensor([  [[1,1], [2, 2]], 
                        [[2,2], [3, 3]],
                        [[4,4], [5, 5]],
                    ], dtype=torch.float32, requires_grad=True)
    print(x.shape)
    S = torch.sparse_coo_tensor(indices=indices, values=x, size=(10, 10, 2, 2), device='cuda')
    print(S)

    f = torch.sparse.sum(S) # calls to_dense(), allocating gigabytes of memory
    print("sum: ", f)
    f.backward()

    # print(S.grad)
    print(x.grad)
```
which are part of the test of `test_sparse_sum`

Reply:
This would work only if values tensor is contiguous - is it guaranteed?

Reply:
Yes it does, I added some TORCH_CHECKs at the beginning of the function.